mne/time_frequency/psd.py

```diff
@@ -146,15 +151,30 @@ def compute_epochs_psd(epochs, picks=None, fmin=0, fmax=np.inf, n_fft=256,
         picks = pick_types(epochs.info, meg=True, eeg=True, ref_meg=False,
                            exclude='bads')
 
+    data = epochs.get_data()[:, picks]
+    if proj:
+        proj, _ = make_projector_info(raw.info)
+        if picks is not None:
+            data = np.dot(proj[picks][:, picks], data)
+        else:
+            data = np.dot(proj, data)
+
     logger.info("Effective window size : %0.3f (s)" % (n_fft / float(Fs)))
-    psds = []
-    import matplotlib.pyplot as plt
-    parallel, my_psd, n_jobs = parallel_func(_compute_psd, n_jobs)
-    fig = plt.figure()  # threading will induce errors otherwise
-    out = parallel(my_psd(data[picks], fmin, fmax, Fs, n_fft, plt.psd,
-                          n_overlap, pad_to)
-                   for data in epochs)
-    plt.close(fig)
-    psds = [o[0] for o in out]
-    freqs = [o[1] for o in out]
-    return np.array(psds), freqs[0]
+
+    freqs = np.arange(n_fft // 2 + 1) * (Fs / n_fft)
+    freq_mask = (freqs >= fmin) & (freqs <= fmax)
+    freqs = freqs[freq_mask]
+    psds = np.empty(data.shape[:-1] + (freqs.size,))
+
+    parallel, my_pwelch, n_jobs = parallel_func(_pwelch, n_jobs=n_jobs,
+                                                verbose=verbose)
+
+    for idx, fepoch in zip(np.array_split(np.arange(len(data)), n_jobs),
+                           parallel(my_pwelch(epoch, window_size=window_size,
+                                    noverlap=n_overlap, nfft=n_fft, fs=Fs,
+                                    freq_mask=freq_mask)
+                           for epoch in np.array_split(data, n_jobs))):
+        for k, p in zip(idx, fepoch):
+            psds[k, :, :] = p
```

Comment:
can we use more explicit identifiers here? i_epochs, epochs... or i_channl, f_channel, etc.